diff --git a/tvtk/pyface/ui/qt4/actor_editor.py b/tvtk/pyface/ui/qt4/actor_editor.py
index e64aa7cb..688c6255 100644
--- a/tvtk/pyface/ui/qt4/actor_editor.py
+++ b/tvtk/pyface/ui/qt4/actor_editor.py
@@ -12,7 +12,10 @@ from pyface.qt import QtGui

 # Enthought library imports.
 from traits.api import Any, Bool, Callable, Dict, Str
-from traitsui.qt4.editor import Editor
+try:
+    from traitsui.qt4.editor import Editor
+except ModuleNotFoundError:
+    from traitsui.qt.editor import Editor
 from traitsui.basic_editor_factory import BasicEditorFactory

 from .decorated_scene import DecoratedScene
diff --git a/tvtk/pyface/ui/qt4/scene_editor.py b/tvtk/pyface/ui/qt4/scene_editor.py
index 00977285..c3b1aad9 100644
--- a/tvtk/pyface/ui/qt4/scene_editor.py
+++ b/tvtk/pyface/ui/qt4/scene_editor.py
@@ -13,7 +13,10 @@ from pyface.qt import QtGui

 # Enthought library imports.
 from traits.api import Any, Bool, Callable
-from traitsui.qt4.editor import Editor
+try:
+    from traitsui.qt.editor import Editor
+except ModuleNotFoundError:  # traitsui < 8
+    from traitsui.qt4.editor import Editor
 from traitsui.basic_editor_factory import BasicEditorFactory

 from .decorated_scene import DecoratedScene
